Nombre: Caroline Michelle Sierra Linares

# **1. Elección del conjunto de datos**

Para iniciar el análisis, se llevó a cabo la instalación de los paquetes necesarios y la carga de herramientas para la manipulación, exploración y visualización de datos. A continuación, se enumeran las librerías utilizadas:

- **dlookr:** Para la generación de reportes descriptivos.  
- **arrow:** Para el manejo de datos en diferentes formatos.  
- **Hmisc** y **psych:** Utilizados para funciones de descripción estadística, como `describe`.  
- **tidyverse:** Un conjunto de paquetes integrados para la manipulación y visualización de datos.  

In [6]:
library(tidyverse) # colecciones de librerias

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.1     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.1
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ ggplot2::%+%()     masks psych::%+%()
✖ ggplot2::alpha()   masks psych::alpha()
✖ tidyr::extract()   masks dlookr::extract()
✖ dplyr::filter()    masks stats::filter()
✖ dplyr::lag()       masks stats::lag()
✖ dplyr::src()       masks Hmisc::src()
✖ dplyr::summarize() masks Hmisc::summarize()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors


In [1]:
install.packages("dlookr")
library("dlookr")

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

Registered S3 methods overwritten by 'dlookr':
  method          from  
  plot.transform  scales
  print.transform scales


Attaching package: ‘dlookr’


The following object is masked from ‘package:base’:

    transform




In [2]:
install.packages("arrow")  

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [1]:
library("arrow") #Libreria para leer paquetes


Attaching package: ‘arrow’


The following object is masked from ‘package:utils’:

    timestamp




In [2]:
df <- read.csv("/kaggle/input/student-performance-dataset/Student_performance_data _.csv") # Creamos el dataframe del datase

# **2. Limpieza y Transformación de los Datos**

En esta fase, realizaremos una exploración detallada del conjunto de datos. Con la inspección tanto de las primeras como de las últimas filas del DataFrame y más información necesaria. Este paso es fundamental para obtener una visión general de la estructura de los datos, identificar posibles inconsistencias y verificar que las columnas estén correctamente formateadas.

Además de explorar las primeras y últimas filas, también realizaremos un análisis de la información general del conjunto de datos, como el tipo de datos de cada columna, la presencia de valores nulos y cualquier valor atípico que pueda requerir limpieza o transformación. Esto nos permitirá determinar qué transformaciones son necesarias para preparar los datos para un análisis más profundo y garantizar que sean adecuados para el modelado y la visualización.

In [1]:
install.packages("Hmisc")
library(Hmisc)
install.packages("psych")
library(psych)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)


Attaching package: ‘Hmisc’


The following objects are masked from ‘package:base’:

    format.pval, units


Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)


Attaching package: ‘psych’


The following object is masked from ‘package:Hmisc’:

    describe




In [5]:
describe(df)

vars n    mean         sd          median      trimmed     
StudentID          1   2392 2196.5000000 690.6552444 2196.500000 2.196500e+03
Age                2   2392   16.4686455   1.1237984   16.000000 1.646082e+01
Gender             3   2392    0.5108696   0.4999864    1.000000 5.135841e-01
Ethnicity          4   2392    0.8775084   1.0284758    0.000000 7.309300e-01
ParentalEducation  5   2392    1.7462375   1.0004111    2.000000 1.745037e+00
StudyTimeWeekly    6   2392    9.7719919   5.6527742    9.705363 9.733088e+00
Absences           7   2392   14.5413880   8.4674174   15.000000 1.457315e+01
Tutoring           8   2392    0.3014214   0.4589712    0.000000 2.518286e-01
ParentalSupport    9   2392    2.1220736   1.1228129    2.000000 2.138454e+00
Extracurricular   10   2392    0.3833612   0.4863068    0.000000 3.542320e-01
Sports            11   2392    0.3035117   0.4598704    0.000000 2.544410e-01
Music             12   2392    0.1969064   0.3977441    0.000000 1.212121e-01
Volunteering      13   2392    0.1571906   0.3640565    0.000000 7.157785e-02
GPA               14   2392    1.9061863   0.9151558    1.893393 1.904202e+00
GradeClass        15   2392    2.9836957   1.2339076    4.000000 3.160397e+00
                  mad        min          max        range      skew       
StudentID         886.594800 1.001000e+03 3392.00000 2391.00000  0.00000000
Age                 1.482600 1.500000e+01   18.00000    3.00000  0.03574855
Gender              0.000000 0.000000e+00    1.00000    1.00000 -0.04346127
Ethnicity           0.000000 0.000000e+00    3.00000    3.00000  0.75811560
ParentalEducation   1.482600 0.000000e+00    4.00000    4.00000  0.21537957
StudyTimeWeekly     6.969028 1.056539e-03   19.97809   19.97704  0.05000530
Absences           10.378200 0.000000e+00   29.00000   29.00000 -0.02598447
Tutoring            0.000000 0.000000e+00    1.00000    1.00000  0.86495866
ParentalSupport     1.482600 0.000000e+00    4.00000    4.00000 -0.16664951
Extracurricular     0.000000 0.000000e+00    1.00000    1.00000  0.47949173
Sports              0.000000 0.000000e+00    1.00000    1.00000  0.85418049
Music               0.000000 0.000000e+00    1.00000    1.00000  1.52342637
Volunteering        0.000000 0.000000e+00    1.00000    1.00000  1.88248820
GPA                 1.072985 0.000000e+00    4.00000    4.00000  0.01450739
GradeClass          0.000000 0.000000e+00    4.00000    4.00000 -0.89967778
                  kurtosis   se          
StudentID         -1.2015051 14.121496616
Age               -1.3731015  0.022977767
Gender            -1.9989463  0.010222981
Ethnicity         -0.7703030  0.021028751
ParentalEducation -0.2903732  0.020454925
StudyTimeWeekly   -1.1373753  0.115579564
Absences          -1.1778001  0.173129223
Tutoring          -1.2523696  0.009384365
ParentalSupport   -0.7254322  0.022957616
Extracurricular   -1.7708275  0.009943281
Sports            -1.2709065  0.009402749
Music              0.3209626  0.008132483
Volunteering       1.5444080  0.007443689
GPA               -0.8657664  0.018711752
GradeClass        -0.4204059  0.025229116

In [3]:
print("Explorando las primeras filas del archivo:")
head(df) # .head() Para ver las primeras 6 filas del archivo en R

[1] "Explorando las primeras filas del archivo:"


StudentID Age Gender Ethnicity ParentalEducation StudyTimeWeekly Absences
1 1001      17  1      0         2                 19.833723        7      
2 1002      18  0      0         1                 15.408756        0      
3 1003      15  0      2         3                  4.210570       26      
4 1004      17  1      0         3                 10.028829       14      
5 1005      17  1      0         2                  4.672495       17      
6 1006      18  0      0         1                  8.191219        0      
  Tutoring ParentalSupport Extracurricular Sports Music Volunteering GPA      
1 1        2               0               0      1     0            2.9291956
2 0        1               0               0      0     0            3.0429148
3 0        2               0               0      0     0            0.1126023
4 0        3               1               0      0     0            2.0542181
5 1        3               0               0      0     0            1.2880612
6 0        1               1               0      0     0            3.0841836
  GradeClass
1 2         
2 1         
3 4         
4 3         
5 4         
6 1

In [6]:
print("Explorando las ultimas filas del archivo:")
tail(df) #Para ver las ultimas 5 filas del archivo

[1] "Explorando las ultimas filas del archivo:"


StudentID Age Gender Ethnicity ParentalEducation StudyTimeWeekly Absences
2387 3387      16  0      0         2                 13.814021       14      
2388 3388      18  1      0         3                 10.680555        2      
2389 3389      17  0      0         1                  7.583217        4      
2390 3390      16  1      0         2                  6.805500       20      
2391 3391      16  1      1         0                 12.416653       17      
2392 3392      16  1      0         2                 17.819907       13      
     Tutoring ParentalSupport Extracurricular Sports Music Volunteering
2387 0        2               0               0      0     0           
2388 0        4               1               0      0     0           
2389 1        4               0               1      0     0           
2390 0        2               0               0      0     1           
2391 0        2               0               1      1     0           
2392 0        2               0               0      0     1           
     GPA      GradeClass
2387 2.000084 4         
2388 3.455509 0         
2389 3.279150 4         
2390 1.142333 2         
2391 1.803297 1         
2392 2.140014 1

In [8]:
# Ver la estructura del dataset, obervar tipos de datos
str(df)

'data.frame':	2392 obs. of  15 variables:
 $ StudentID        : int  1001 1002 1003 1004 1005 1006 1007 1008 1009 1010 ...
 $ Age              : int  17 18 15 17 17 18 15 15 17 16 ...
 $ Gender           : int  1 0 0 1 1 0 0 1 0 1 ...
 $ Ethnicity        : int  0 0 2 0 0 0 1 1 0 0 ...
 $ ParentalEducation: int  2 1 3 3 2 1 1 4 0 1 ...
 $ StudyTimeWeekly  : num  19.83 15.41 4.21 10.03 4.67 ...
 $ Absences         : int  7 0 26 14 17 0 10 22 1 0 ...
 $ Tutoring         : int  1 0 0 0 1 0 0 1 0 0 ...
 $ ParentalSupport  : int  2 1 2 3 3 1 3 1 2 3 ...
 $ Extracurricular  : int  0 0 0 1 0 1 0 1 0 1 ...
 $ Sports           : int  0 0 0 0 0 0 1 0 1 0 ...
 $ Music            : int  1 0 0 0 0 0 0 0 0 0 ...
 $ Volunteering     : int  0 0 0 0 0 0 0 0 1 0 ...
 $ GPA              : num  2.929 3.043 0.113 2.054 1.288 ...
 $ GradeClass       : num  2 1 4 3 4 1 2 4 2 0 ...



Como se observa en la tabla, unciamente son números para todas las variables, por lo que para continuar con el análisis, se investigó el significado de cada variable, el cual en la pagina de Kaggle nos dice lo siguiente: 

**Datos demográficos**

*   **Género :** Género de los estudiantes, donde 0 representa masculino y 1 representa femenino.

* **Etnicidad :** La etnicidad de los estudiantes, codificada de la siguiente manera:
  *  0: caucásico
  * 1: Afroamericano
  * 2: asiático
  * 3: Otros

* Educación de los padres : El nivel de educación de los padres, codificado de la siguiente manera:
  * 0: Ninguno
  * 1: Escuela secundaria
  * 2: Un poco de universidad
  * 3: Licenciatura
  * 4: Superior

**Hábitos de estudio**

* **StudyTimeWeekly** : Tiempo de estudio semanal en horas, de 0 a 20.

* **Ausencias** : Número de ausencias durante el año escolar, comprendido entre 0 y 30.

* **Tutoría** : Estado de la tutoría, donde 0 indica No y 1 indica Sí.

**Participación de los padres**

* **ParentalSupport** : El nivel de apoyo de los padres, codificado de la siguiente manera:
  * 0: Ninguno
  * 1: Bajo
  * 2: Moderado
  * 3: Alto
  * 4: Muy alto

**Actividades extracurriculares**

* **Extracurricular** : Participación en actividades extracurriculares, donde 0 indica No y 1 indica Sí.

* **Deportes** : Participación en deportes, donde 0 indica No y 1 indica Sí.

* **Música** : Participación en actividades musicales, donde 0 indica No y 1 indica Sí.

* **Voluntariado** : Participación en voluntariado, donde 0 indica No y 1 indica Sí.

**Rendimiento académico**

* **GPA** : Promedio de calificaciones en una escala de 2.0 a 4.0, influenciado por los hábitos de estudio, la participación de los padres y las actividades extracurriculares.

**Variable objetivo**: Grado Clase 
* **GradeClass**: Clasificación de las calificaciones de los estudiantes según el GPA:
  * **0**: 'A' (GPA >= 3.5)
  * **1**: 'B' (3.0 <= GPA < 3.5)
  * **2**: 'C' (2,5 <= GPA < 3,0)
  * **3**: 'D' (2.0 <= GPA < 2.5)
  * **4**: 'F' (GPA < 2.0)Cx

## **3. Análisis Exploratorio de Datos (EDA) y Visualización de Resultados**

Ahora, se procede a realizar un **Análisis Exploratorio de Datos (EDA)**. Para ello, se utilizó una herramienta que permite generar un reporte EDA automático, facilitando la exploración y comprensión de los datos de manera rápida y eficiente.

Para utilizar esta herramienta, instalaremos la librería **`dlookr`**


In [7]:
install.packages("dlookr")
library("dlookr")

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

Registered S3 methods overwritten by 'dlookr':
  method          from  
  plot.transform  scales
  print.transform scales


Attaching package: ‘dlookr’


The following object is masked from ‘package:psych’:

    describe


The following object is masked from ‘package:Hmisc’:

    describe


The following object is masked from ‘package:base’:

    transform




**Dlookr tiene una función llamada `eda_web_report()` que permite generar un reporte automático del análisis exploratorio de datos (EDA). A continuación se detallan los parámetros utilizados en esta función:**

- `target`: Representa la variable de salida a estudiar. En este caso, hemos seleccionado **GradeClass** como la variable objetivo, ya que clasifica el rendimiento académico de los estudiantes según su GPA.
  
- `output_file`: Especifica el nombre del archivo del reporte cuando se descargue. En este ejemplo, el archivo se llamará **reporte_EDA_con_R.html**.

- `output_dir`: Indica el directorio de salida donde se guardará el reporte generado. Se utiliza la función `getwd()` para obtener el directorio de trabajo actual.

- `titel`: Define el título del reporte. En este caso, es **"Reporte EDA-R"**.

- `subtitle`: Especifica el subtítulo del reporte. En este ejemplo, se ha utilizado **"Actividad con R"**.

- `theme`: Permite personalizar el tema del reporte. En este caso, se ha elegido el tema **"blue"**.

Este conjunto de parámetros genera un reporte interactivo que facilita la comprensión y visualización de los datos, lo cual es útil para un análisis más profundo y detallado.


Elegí como variable objetivo la columna GradeClass, ya que esta columna clasifica las calificaciones de los estudiantes en función de su GPA, lo que la convierte en una medida clave del rendimiento académico. Esta variable es influenciada por factores como el tiempo de estudio, la participación en actividades extracurriculares, el apoyo de los padres, entre otros. Por lo tanto, GradeClass representa una forma efectiva de categorizar el rendimiento académico de los estudiantes y entender cómo los diferentes factores impactan en su desempeño.

In [13]:
eda_web_report(df,
               target="GradeClass",
               output_file="reporte_EDA_con_R.html",
               output_dir= getwd(),
                titel= "Reporte EDA-R",
                subtitle= "Actividad con R",
                theme="blue"
              )



processing file: eda_temp.Rmd



1/63                    
2/63 [setup]            
3/63                    
4/63 [load_packages]    
5/63                    
6/63 [unnamed-chunk-1]  
7/63                    
8/63 [udf]              
9/63                    
10/63 [check_variables]  
11/63                    
12/63 [create-overview]  
13/63                    
14/63 [overview]         
15/63                    
16/63 [overview-pre]     
17/63                    
18/63 [unnamed-chunk-2]  
19/63                    
20/63 [unnamed-chunk-3]  
21/63                    
22/63 [variables]        
23/63                    
24/63 [normality]        
25/63                    
26/63 [normality-list]   
27/63                    
28/63 [unnamed-chunk-4]  
29/63                    
30/63 [unnamed-chunk-5]  
31/63                    
32/63 [compare_numerical]
33/63                    
34/63 [unnamed-chunk-6]  
35/63                    
36/63 [compare-category] 
37/63                    
38/63 [unnamed-chunk-7]  
39/63                

output file: eda_temp.knit.md




/usr/bin/pandoc +RTS -K512m -RTS eda_temp.knit.md --to html4 --from markdown+autolink_bare_uris+tex_math_single_backslash --output /kaggle/working/reporte_EDA_con_R.html --lua-filter /usr/local/lib/R/site-library/rmarkdown/rmarkdown/lua/pagebreak.lua --lua-filter /usr/local/lib/R/site-library/rmarkdown/rmarkdown/lua/latex-div.lua --embed-resources --standalone --variable bs3=TRUE --section-divs --template /usr/local/lib/R/site-library/rmarkdown/rmd/h/default.html --no-highlight --variable highlightjs=1 --variable theme=bootstrap --css /usr/local/lib/R/site-library/dlookr/resources/dlookr-bootstrap.css --mathjax --variable 'mathjax-url=https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML' --include-in-header /tmp/Rtmpc1zuLN/rmarkdown-str216eaef8d3.html --variable code_folding=show --variable code_menu=1 --include-in-header header_temp.html --include-after-body /usr/local/lib/R/site-library/dlookr/resources/footer.html 



Output created: reporte_EDA_con_R.html



Según el reporte, no se encuentran datos nulos, pero verificaremos si hay filas o datos duplicados.

In [10]:
# Verificar si hay filas duplicadas
duplicated_rows <- duplicated(df)
sum(duplicated_rows)  # Cuenta cuántas filas están duplicadas

[1] 0

Ahora verificaremos únicamente si hay duplicados en la columna `StudentID`, ya que es el único caso en el que tendría sentido eliminar los registros duplicados, ya que los demás son datos numéricos que pueden repetirse entre diferentes estudiantes.

In [12]:
# Verificar duplicados solo por 'StudentID'
duplicated_ids <- duplicated(df$StudentID)
sum(duplicated_ids)  # Cuenta cuántos duplicados hay en 'StudentID'

[1] 0

Al no tener datos duplicados ni nulos, procedemos a rescatar lo que se observó y se enfocó en el informe generado previamente.

# **4. Conclusiones y hallazgos importantes**
# **Conclusiones Generales**

1. **Influencia significativa del GPA en GradeClass:**  
   Se observó que el **GPA** tiene una fuerte correlación negativa con **GradeClass**, indicando que a medida que el rendimiento académico mejora (mayor **GPA**), las calificaciones de los estudiantes también mejoran. Esto refuerza la idea de que el **GPA** es un predictor clave para clasificar el rendimiento académico y debe ser considerado como una variable principal en cualquier modelo predictivo.

2. **Importancia de la asistencia a clases:**  
   La correlación positiva significativa entre **Ausencias** y **GradeClass** demuestra que los estudiantes con mayor cantidad de ausencias tienden a obtener calificaciones más bajas. Esto subraya la necesidad de incluir la variable **Ausencias** como un factor crítico en los modelos de predicción, ya que su impacto en el rendimiento académico es considerable.

3. **Relevancia limitada del tiempo de estudio semanal:**  
   Aunque la relación entre **StudyTimeWeekly** y **GradeClass** es débil, no debe descartarse por completo. Es posible que esta variable tenga un impacto más significativo cuando se analiza en combinación con otras características, como el **GPA** o las **Ausencias**.

4. **Impacto indirecto del apoyo parental:**  
   La débil correlación negativa entre **ParentalSupport** y **GradeClass** sugiere que, aunque el apoyo de los padres no tiene un impacto directo fuerte en las calificaciones, podría influir indirectamente en otras variables, como los hábitos de estudio o la asistencia.


## **Correlaciones detectadas**

1. **GradeClass y GPA (-0.78):**  
   Se detecta una fuerte correlación negativa entre **GradeClass** y **GPA**. Esto indica que a medida que el **GPA** aumenta, GradeClass mejora, y viceversa. Un **GPA** bajo (menos de 2.0) está asociado con una **GradeClass** baja, es decir una calificación "F", mientras que un **GPA** alto (más de 3.5) está relacionado con una **GradeClass** alta, calificación "A".

   También, marcada en color rojo puesto que si estan relacionadas de manera directa, ya  que una varoable influye con la otra, pues **GradeClass** depende del **GPA**.

3. **GradeClass y Ausencias (+0.73):**  
   Existe una fuerte correlación positiva entre **GradeClass** y **Ausencias**. Esto puede estar relacionado a que los estudiantes que tienen más ausencias en clase tienden a obtener calificaciones más bajas. Por lo que podemos decir que la asistencia a clases es un factor importante para el rendimiento académico.

4. **GradeClass y StudyTimeWeekly (-0.13):**  
   La correlación negativa débil entre **GradeClass** y **StudyTimeWeekly** sugiere que el tiempo dedicado al estudio semanal no tiene una relación fuerte con las calificaciones. Esto podría indicar que otros factores son más relevantes que simplemente la cantidad de tiempo invertido.

5. **GradeClass y ParentalSupport (-0.14):**  
   La correlación negativa débil entre **GradeClass** y **ParentalSupport** indica que el apoyo de los padres tiene una influencia limitada en las calificaciones de los estudiantes en este caso. Ya que el impacto directo en el rendimiento académico parece ser menos significativo en comparación con otros factores que se estudian.

___
# **Próximos pasos a seguir**

-  **Explorar la influencia de las ausencias en el rendimiento académico:**  
   La fuerte correlación entre **GradeClass** y **Ausencias** implica que las ausencias tienen un impacto significativo en el rendimiento académico. Se debe considerar la **ausencia** como una variable importante en el modelo, porque esto puede incluir técnicas de ingeniería de características para capturar patrones de ausencias y cómo afectan las calificaciones.

-  **Revisar la relación débil entre StudyTimeWeekly y GradeClass:**  
   Aunque la correlación entre **StudyTimeWeekly** y **GradeClass** es débil, es importante no descartarlo por completo. Es posible que la cantidad de tiempo de estudio no sea un predictor directo, pero puede ser útil cuando se combina con otras variables.

-  **Probar diferentes algoritmos de Machine Learning** 

-  **Ajustar el impacto del apoyo parental en el modelo:**  
   Aunque **ParentalSupport** muestra una correlación débil con **GradeClass**, podría ser útil incluirlo en el modelo, ya que en algunos contextos podría tener un impacto indirecto.